runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java

```diff
@@ -307,25 +310,89 @@ private void calculate()
 	{
 		for (UIActionSlot slot : uiActionSlots)
 		{
-			int actionCount = 0;
-			int neededXP = targetXP - currentXP;
-			SkillDataEntry action = slot.getAction();
-			double xp = (action.isIgnoreBonus()) ? action.getXp() : action.getXp() * xpFactor;
+			final SkillDataEntry action = slot.getAction();
+			final ActionSequenceInfo sequence = calculateActionSequence(action, currentXP, targetXP, xpFactor);
 
-			if (neededXP > 0)
-			{
-				actionCount = (int) Math.ceil(neededXP / xp);
-			}
-
-			slot.setText("Lvl. " + action.getLevel() + " (" + formatXPActionString(xp, actionCount, "exp) - "));
+			slot.setText("Lvl. " + action.getLevel() + " (" + formatXPActionString(sequence.getActionXp(), sequence.getActionCount(), "exp) - "));
 			slot.setAvailable(currentLevel >= action.getLevel());
 			slot.setOverlapping(action.getLevel() < targetLevel);
-			slot.setValue(xp);
+			slot.setValue(sequence.getActionXp());
 		}
 
 		updateCombinedAction();
 	}
 
+	@VisibleForTesting
+	static ActionSequenceInfo calculateActionSequence(SkillDataEntry action, int startXP, int goalXP, float xpFactor)
+	{
+		final int neededXp = goalXP - startXP;
+		double actionXp = action.isIgnoreBonus() ? action.getXp() : action.getXp() * xpFactor;
+
+		if (neededXp <= 0)
+		{
+			return new ActionSequenceInfo(actionXp, 0);
+		}
+
+		if (action.getXpScalings() == null)
+		{
+			return new ActionSequenceInfo(actionXp, (int) Math.ceil(neededXp / actionXp));
+		}
+
+		int accumulatedActions = 0;
+		double accumulatedXp = 0;
+
+		while (accumulatedXp < neededXp)
+		{
+			final int accumulatedLvl = Experience.getLevelForXp((int) (startXP + accumulatedXp));
+			actionXp = calculateScaledActionXp(accumulatedLvl, action);
+			actionXp *= action.isIgnoreBonus() ? 1 : xpFactor;
```

Comment:
Is this applying `xpFactor` a second time to `actionXp`? This exact calculation is already done at the top of this method.